arborio/cableio.cpp

```diff
@@ -0,0 +1,799 @@
+place_pair make_place(locset where, placeable what) {
+    return place_pair{where, what};
+}
+paint_pair make_paint(region where, paintable what) {
+    return paint_pair{where, what};
+}
+defaultable make_default(defaultable what) {
+    return what;
+}
+decor make_decor(const std::vector<std::variant<place_pair, paint_pair, defaultable>>& args) {
+    decor d;
+    for(const auto& a: args) {
+        auto decor_visitor = arb::util::overload(
+            [&](const place_pair & p) { d.place(p.first, p.second); },
+            [&](const paint_pair & p) { d.paint(p.first, p.second); },
+            [&](const defaultable & p){ d.set_default(p); });
+        std::visit(decor_visitor, a);
+    }
+    return d;
+}
+
+// Define maker for locset pairs, region pairs and label_dicts
+using locset_pair = std::pair<std::string, locset>;
+using region_pair = std::pair<std::string, region>;
+locset_pair make_locset_pair(std::string name, locset desc) {
+    return locset_pair{name, desc};
+}
+region_pair make_region_pair(std::string name, region desc) {
+    return region_pair{name, desc};
+}
+label_dict make_label_dict(const std::vector<std::variant<locset_pair, region_pair>>& args) {
+    label_dict d;
+    for(const auto& a: args) {
+        std::visit([&](auto&& x){d.set(x.first, x.second);}, a);
+    }
+    return d;
+}
+// Define makers for mpoints and msegments and morphologies
+using branch = std::tuple<int, int, std::vector<arb::msegment>>;
+arb::mpoint make_point(double x, double y, double z, double r) {
+    return arb::mpoint{x, y, z, r};
+}
+arb::msegment make_segment(unsigned id, arb::mpoint prox, arb::mpoint dist, int tag) {
+    return arb::msegment{id, prox, dist, tag};
+}
+morphology make_morphology(const std::vector<std::variant<branch>>& args) {
+    segment_tree tree;
+    std::vector<unsigned> branch_final_seg(args.size());
+    std::vector<unsigned> branch_children(args.size(), 0);
+    std::vector<std::pair<msegment, int>> segs;
+    for (const auto& br: args) {
+        auto b = std::get<branch>(br);
+        auto b_id = std::get<0>(b);
+        auto b_pid = std::get<1>(b);
+        auto b_segments = std::get<2>(b);
+
+        if (b_pid!=-1) branch_children[b_pid]++;
+
+        auto s_pid = b_pid==-1? arb::mnpos: branch_final_seg[b_pid];
+        for (const auto& s: b_segments) {
+            segs.emplace_back(s, s_pid);
+            s_pid = s.id;
+        }
+        branch_final_seg[b_id] = s_pid;
+    }
+    // A branch should have either 0 or >1 children.
+    auto it = std::find_if(branch_children.begin(), branch_children.end(), [](unsigned i){return i==1;});
+    if (it != branch_children.end()) {
+        throw cableio_morphology_error(it - branch_children.begin());
+    }
+
+    // Append segments according to id order.
+    std::sort(segs.begin(), segs.end(), [](const auto& lhs, const auto& rhs){return lhs.first.id < rhs.first.id;});
+    for (const auto& spair: segs) {
```

Comment:
You know what I am going to suggest here, right? ;)

`const auto& [seg, s_pid]`

Reply:
Right, still not used to that. I'll make the change, thanks!